02-express-tutorial/answers/11-methods.js

```diff
@@ -0,0 +1,80 @@
+const express = require('express');
+const app = express();
+require('dotenv').config();
+const port = process.env.PORT || 3000;
+const { people } = require('./data');
+
+//static assets
+app.use(express.static('./methods-public'));
+// parse form data
+app.use(express.urlencoded({ extended: false }));
+
+// parse.json
+app.use(express.json());
+
+app.post('/login', (req, res) => {
+  const { name } = req.body;
+  if (name) {
+    return res.status(200).send(`Welcome, ${name}!`);
+  } else {
+    return res.status(401).send('Please provide credentials');
+  }
+});
+
+app.get('/api/people', (req, res) => {
+  res.status(200).json({ success: true, data: people });
+});
+
+app.post('/api/people', (req, res) => {
+  const { name } = req.body;
+  if (!name) {
+    return res
+      .status(400)
+      .json({ success: false, msg: 'please provide a name' });
+  }
+  res.status(201).json({ success: true, person: name });
+});
+
+app.post('/api/postman/people', (req, res) => {
+  const { name } = req.body;
+  if (!name) {
+    return res
+      .status(400)
+      .json({ success: false, msg: 'please provide a name' });
+  }
+  res.status(201).json({ success: true, data: [...people, name] });
+});
+
+app.put('/api/people/:id', (req, res) => {
+  const { id } = req.params;
+  const { name } = req.body;
+  const person = people.find((person) => person.id === Number(id));
+  if (!person) {
+    return res
+      .status(404)
+      .json({ success: false, msg: `person not found with this ${id}` });
+  }
+  const newPeople = people.map((person) => {
+    if (person.id === Number(id)) {
+      person.name = name;
+    }
+    return person;
+  });
+  res.status(200).json({ success: true, data: newPeople });
```

Comment:
Question: why can we sometimes `return` the result of res.status().send(), and other times _not return_ the same thing and it works the same either way? When do we need to return? Is there anything that is important to avoid with `res.send()`?

Reply:
We use ```return``` when we want to immediately send a response back to the client and terminate the current request-response cycle. Return will also ensure that the response is sent and the server doesn't continue to execute any remaining code that might be associated with the current route handler. In some cases, we don't need to return the result such as when we perform asynchronous operations. Also, we need to avoid calling ```res.send()``` multiple times in the same route handler because it can cause errors.